src/systems/gimbal_controller/GimbalControllerPlugin.cc

```diff
@@ -0,0 +1,484 @@
+}
+
+//////////////////////////////////////////////////
+void GimbalControllerPlugin::PreUpdate(const ignition::gazebo::UpdateInfo &_info,
+    ignition::gazebo::EntityComponentManager &_ecm)
+{  if(!this->dataPtr->imuInitialized)
+   {
+      // Set unconditionally because we're only going to try this once.
+      this->dataPtr->imuInitialized = true;
+
+      auto imuEntity = _ecm.EntityByComponents(
+          components::Name(this->dataPtr->imuName),
+          components::Imu(),
+          components::ParentEntity(this->dataPtr->modelLink));
+
+      // Imu sensor topic 
+      std::string topic = scopedName(imuEntity, _ecm) + "/imu";
+
+      if(topic.empty())
+      {
+        ignerr << "imu_sensor [" << this->dataPtr->imuName
+               << "] not found, abort ArduPilot plugin." << "\n";
+        return;
+      }
+
+      // Subscribe to IMU sensor topic
+      this->dataPtr->node.Subscribe(topic, 
+        &GimbalControllerPluginPrivate::imuCb);
+    }
+    
+    if(!this->dataPtr->pitchJoint || !this->dataPtr->rollJoint 
+        || !this->dataPtr->yawJoint)
+      return;
+
+    if(_info.simTime < this->dataPtr->lastControllerUpdateTime)
+    {    
+       ignerr << "time reset event\n";
+       this->dataPtr->lastControllerUpdateTime = _info.simTime;
+        return;
+    }
+    else if (_info.simTime > this->dataPtr->lastControllerUpdateTime)
+    { 
+      // Time delta
+      std::chrono::duration<double> dt  = 
+      (this->dataPtr->lastControllerUpdateTime - _info.simTime);
+  
+      // Compute pitch joint's position and set forces
+      auto pitchJointPos = _ecm.Component<components::JointPosition>
+      (this->dataPtr->pitchJoint);
+      
+      ignition::gazebo::components::JointForceCmd* pjfcComp = nullptr;
+      pjfcComp = _ecm.Component<components::JointForceCmd>
+      (this->dataPtr->pitchJoint);
+
+      if (pjfcComp == nullptr)
+      {
+        pjfcComp = _ecm.Component<components::JointForceCmd>
+        (_ecm.CreateComponent(this->dataPtr->pitchJoint, 
+          components::JointForceCmd({0})));
+      }
+
+      // Compute yaw joint's position and set forces
+      auto yawJointPos = _ecm.Component<components::JointPosition>
+      (this->dataPtr->yawJoint);
+      
+      ignition::gazebo::components::JointForceCmd* yjfcComp = nullptr;
+      yjfcComp = _ecm.Component<components::JointForceCmd>
+      (this->dataPtr->yawJoint);
+
+      if (yjfcComp == nullptr)
+      {
+        yjfcComp = _ecm.Component<components::JointForceCmd>
+        (_ecm.CreateComponent(this->dataPtr->yawJoint, 
+          components::JointForceCmd({0})));
+      }
+
+      // Compute roll joint's position and set forces
+      auto rollJointPos = _ecm.Component<components::JointPosition>
+      (this->dataPtr->rollJoint);
+
+      ignition::gazebo::components::JointForceCmd* rjfcComp = nullptr;
+      rjfcComp = _ecm.Component<components::JointForceCmd>
+      (this->dataPtr->rollJoint);
+
+      if (rjfcComp == nullptr)
```

Comment:
This is only setting the command if the component doesn't exist yet. If it exists, the command won't be set. Is that expected?